AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStoreAdmin/src/InitializeMutation.dfy

```diff
@@ -0,0 +1,612 @@
+    var timestamp :- timestamp?
+    .MapFailure(e => Types.KeyStoreAdminException(
+                    message := "Could not generate a timestamp: " + e));
+
+    var mutationCommitmentUUID? := UUID.GenerateUUID();
+    var mutationCommitmentUUID :- mutationCommitmentUUID?
+    .MapFailure(e => Types.KeyStoreAdminException(
+                    message := "Could not generate UUID for Mutation Commitment. " + e));
+
+    var inferredOriginalEC
+      := map k <- activeItem.EncryptionContext
+                  // This pull everything that is not in our restricted list.
+             | k !in Structure.BRANCH_KEY_RESTRICTED_FIELD_NAMES
+      :: k := activeItem.EncryptionContext[k];
+
+    // To Preserve Unexpected/un-modeled Attributes.
+    // We need to copy them from inferredOriginalEC to terminalEC.
+    // Which means we need to select those members without a prefix,
+    // and copy them over to terminal.
+    var unexpectedEC := PrefixUtils.FilterMapForKeysThatDoNotBeginWithPrefix(
+      prefix := Structure.ENCRYPTION_CONTEXT_PREFIX,
+      aMap := inferredOriginalEC
+    );
+    assert unexpectedEC.Keys !! Structure.BRANCH_KEY_RESTRICTED_FIELD_NAMES;
+
+    var terminalEC?: Option<KeyStoreTypes.EncryptionContextString> := None;
+    if (input.Mutations.TerminalEncryptionContext.Some?) {
+
+      var terminalEC := PrefixUtils.AddingPrefixToKeysOfMapDoesNotCreateCollisions(
+        prefix := Structure.ENCRYPTION_CONTEXT_PREFIX,
+        aMap := input.Mutations.TerminalEncryptionContext.value
+      ) + unexpectedEC;
+      // ValidateInitializeMutationInput SHOULD take care of this Need, but Dafny is struggling
+      // TODO-Mutations-FF : Replace runtime check with Lemma.
+      // See https://github.com/aws/aws-cryptographic-material-providers-library/pull/750#discussion_r1777654751
+      :- Need(
+        terminalEC.Keys !! Structure.BRANCH_KEY_RESTRICTED_FIELD_NAMES,
+        Types.KeyStoreAdminException(message:="Terminal Encryption Context contains a reserved word!")
+      );
+      terminalEC? := Some(terminalEC);
+      assert terminalEC.Keys !! Structure.BRANCH_KEY_RESTRICTED_FIELD_NAMES;
+    }
+
+    assert KmsArn.ValidKmsArn?(activeItem.KmsArn);
+    var MutationToApply := StateStrucs.MutationToApply(
+      Identifier := input.Identifier,
+      Original := StateStrucs.MutableProperties(
+        kmsArn := activeItem.KmsArn,
+        customEncryptionContext := inferredOriginalEC
+      ),
+      Terminal := StateStrucs.MutableProperties(
+        kmsArn := input.Mutations.TerminalKmsArn.UnwrapOr(activeItem.KmsArn),
+        customEncryptionContext := terminalEC?.UnwrapOr(inferredOriginalEC)
+      ),
+      ExclusiveStartKey := None,
+      UUID := mutationCommitmentUUID,
+      CreateTime := timestamp,
+      Input := input.Mutations,
+      CommitmentCiphertext := [0], // TODO-Mutations-GA Create Commitment Ciphertext
+      IndexCiphertext := [0] // TODO-Mutations-GA Create Index Ciphertext
```

Comment:
is this still supposed to be 0?

Reply:
I am in a rough spot here.
An internal type for Commitment and Index would allow us to not have a Ciphertext blob.
But as long as we are using the datatype's from `Types.dfy`, 
we need to have the Ciphertext field set.

Even though there is no logical value for the ciphertext value here.

I will try and get a refactor for this in before GA on Dec 9th.